### PR TITLE
Add support for worker nodes

### PR DESCRIPTION
**Pull request by ktsakalozos**
The two main updates:
- allow handling worker nodes
- some retries in snap installing and joining nodes

### CODE REVIEW

File: controllers/microk8sconfig_controller.go [hunk 1]
```diff
@@ -335,6 +329,91 @@ func (r *MicroK8sConfigReconciler) handleJoiningControlPlaneNode(ctx context.Con
...
+	if err != nil {
+		scope.Error(err, "Lookup control plane nodes")
+		return ctrl.Result{}, err
+	}
+
+	for _, node := range nodes {
+		if ipOfNodeToConnectTo != "" {
+			break
+		}
+		if node.Spec.ProviderID != nil && node.Status.Phase == "Running" {
```
Comment (lferran) [inline controllers/microk8sconfig_controller.go:381]: Shouldn't only the node that ran `microk8s add-node` be used?

File: controllers/cloudinit/controlplane_init.go [hunk 2]
```diff
@@ -140,6 +140,7 @@ type ControlPlaneInput struct {
 	JoinToken                string
 	ClusterConfiguration     string
 	InitConfiguration        string
+	Version                  string
```
Comment (lferran) [inline controllers/cloudinit/controlplane_init.go:143]: You are not using this one in the template. Is this meant to be the snap channel?

File: controllers/microk8sconfig_controller.go (hunk 1 above)
Comment (ktsakalozos) [reply controllers/microk8sconfig_controller.go:381]: We inject the same join token to all nodes so it does not matter what control plane node we will contact during the join command.

File: controllers/cloudinit/controlplane_init.go (hunk 2 above)
Comment (ktsakalozos) [reply controllers/cloudinit/controlplane_init.go]: You are right the version addition here is not needed.

File: controllers/cloudinit/worker_join.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+// WorkerJoinInput defines context to generate instance user data for worker nodes to join.
+type WorkerJoinInput struct {
+	BaseUserData
+	secret.Certificates
+	JoinToken        string
+	IPOfNodeToJoin   string
+	PortOfNodeToJoin string
+	Version          string
```
Comment (lferran) [inline controllers/cloudinit/worker_join.go:52]: also not used

Comment (ktsakalozos) [reply controllers/cloudinit/worker_join.go:52]: I believe these are used. The controller fills them up and the `generate()` below places them in the cloudinit

**Review by lferran — APPROVED** [review]

**Pull request merged**